### ISSUE TITLE
[RFC] Pipeline lifecycle and state tracking

### ISSUE THREAD

**Issue by ryzhyk** [feldera/feldera#322]
I am refactoring how the pipeline manager tracks and reports the state of the pipeline based on the following principles:

1. The manager determines the state of the pipeline based on its knowledge of its previous observed state and by polling the pipeline's `/stats` endpoint.

2. The manager records the state of the pipeline in the database; this state is returned to the client as part of the pipeline descriptor by the `GET /pipelines` and `GET /pipeline` endpoints.

3. The manager polls pipeline status (1) periodically where the period may vary depending on the state, e.g., it may poll more often when the pipeline is initializing, and (2) immediately after actions that change the state of the pipeline `/start`.

4. All manager endpoints that modify the state of the pipeline (`/deploy`, `/shutodwn`, `/pause`, `/start`) return instantly after forwarding the request to the pipeline, without waiting for the request to complete.  Afterwards, the client may poll the manager for updated status.

Here is the detailed description of the pipeline automaton, which will be in `db/mod.rs`

```rust
/// This type represents the state of the pipeline tracked by the pipeline manager and
/// returned to the API client as part of a [pipeline descriptor](`PipelineDescr`).
///
/// # The lifecycle of a pipeline
/// 
/// The following automaton captures the lifecycle of the pipeline.  Individual states
/// and transitions of the automaton are described below.
///
/// * In addition to the transitions shown in the diagram, all states have an implicit
///   "forced shutdown" transition to the `Shutdown` state.  This transition is triggered
///   when the manager is unable to communicate with the pipeline and forces the pipeline
///   to terminate by sending a `KILL` signal to the pipeline process or by deleting its
///   Kubernetes pod.
///
/// * States labeled with the hourglass symbol (⌛) are **timed** states.  The automaton
///   stays in timed state until the corresponding operation completes or until the manager
///   performs a forced shutdown of the pipeline after a pre-defined timeout perioud.
///
/// * The `Provisioned` hyperstate contains all states where the pipeline's HTTP server is up
///   and running and the manager can query and modify the pipeline's state via its REST
///   endpoints.
///
/// * State transitions labeled with API endpoint names (`/deploy`, `/start`, `/pause`,
///   `/shutdown`) are triggered by invoking corresponding endpoint, e.g.,
///   `POST /v0/pipelines/{pipeline_id}/start`.
///
/// ```text
///                  Shutdown◄────┐
///                     │         │
///                     │         │
///              /deploy│         │
///                     │   ⌛ShuttingDown
///                     ▼         ▲
///             ⌛Provisioning    │
///                     │         │
/// ┌───────────────────┼─────────┼───────────┐
/// │Provisioned        │         │           │
/// │                   ▼         │/shutdown  │
/// │           ⌛Initializing    │           │
/// │                   │         │           │
/// │          ┌────────┴─────────┴─┐         │
/// │          │        ▼           │         │
/// │          │      Paused        │         │
/// │          │      │    ▲        │         │
/// │          │/start│    │/pause  │         │
/// │          │      ▼    │        │         │
/// │          │     Running        │         │
/// │          └──────────┬─────────┘         │
/// │                     │                   │
/// │                     ▼                   │
/// │                   Failed                │
/// └────────────────┬────────────────────────┘
///                  │       ▲
///                  ▼       │
///                 Unreachable
/// ```
#[derive(Serialize, ToSchema, Eq, PartialEq, Debug, Clone, Copy)]
#[cfg_attr(test, derive(proptest_derive::Arbitrary))]
pub(crate) enum PipelineStatus {
    /// Pipeline has not been started or has been shut down.
    ///
    /// The pipeline remains in this state until the user triggers
    /// a deployment by invoking the `/deploy` endpoint.
    Shutdown {
        /// If the previous run of the pipeline failed, contains the cause of the failure.
        error: Option<ErrorResponse>,
    },

    /// The manager triggered a deployment of the pipeline and is
    /// waiting for the pipeline HTTP server to come up.
    ///
    /// In this state, the manager provisions a runtime for the pipeline
    /// (e.g., a Kubernetes pod or a local process), starts the pipeline
    /// within this runtime and waits for it to start accepting HTTP
    /// requests.
    ///
    /// The manager is unable to communicate with the pipeline during this
    /// time.  The pipeline remains in this state until:
    ///
    /// 1. Its HTTP server is up and running; the pipeline transitions
    ///    to the [`Initializing`] state.
    /// 2. A pre-defined timeout has passed.  The manager performs forced
    ///    shutdown of the pipeline;
    ///    returns to the [`Shutdown`] state.
    /// 3. The user cancels the pipeline by invoking the `/shutdown` endpoint.
    ///    The manager performs forced shutdown of the pipeline, returns to
    ///    the [`Shutdown`] state.
    Provisioning,

    /// The pipeline is initializing its internal state and connectors.
    ///
    /// This state is part of the pipeline's provisioning process.  In this state,
    /// the pipeline's HTTP server is up and running, but its query engine
    /// and input and output connectors are still initializing.  The manager is
    /// able to query the status of the pipeline using its status endpoint, but
    /// no data is being processed yet.
    ///
    /// The pipeline remains in this state until:
    ///
    /// 1. Intialization completes successfully; the pipeline transitions
    ///    to the [`Paused`] state.
    /// 2. Intialization fails; transitions to the [`Failed`] state.
    /// 3. A pre-defined timeout has passed.  The manager performs forced shutdown of
    ///    the pipeline; returns to the `Shutdown` state.
    /// 4. The user cancels the pipeline by invoking the `/shutdown` endpoint.
    ///    The manager performs forced shutdown of the pipeline, returns to the
    ///    [`Shutdown`] state.
    /// 5. An unexpected runtime error renders the pipeline [`Unrechable`].
    Initializing,

    /// The pipeline is fully initialized, but data processing has been paused.
    ///
    /// The pipeline remains in this state until:
    ///
    /// 1. The user starts the pipeline by invoking the `/start` endpoint.
    ///    The manager passes the request to the pipeline; transitions to the
    ///    [`Running`] state.
    /// 2. The user cancels the pipeline by invoking the `/shutdown` endpoint.
    ///    The manager passes the shutdown request to the pipeline to perform a graceful
    ///    shutdown; transitions to the [`ShuttingDown`] state.
    /// 3. An unexpected runtime error renders the pipeline [`Unrechable`].
    Paused,

    /// The pipeline is processing data.
    ///
    /// The pipeline remains in this state until:
    ///
    /// 1. The user pauses the pipeline by invoking the `/pause` endpoint.
    ///    The manager passes the request to the pipeline; transitions to the
    ///    [`Paused`] state.
    /// 2. The user cancels the pipeline by invoking the `/shutdown` endpoint.
    ///    The manager passes the shutdown request to the pipeline to perform a graceful
    ///    shutdown; transitions to the [`ShuttingDown`] state.
    /// 3. An unexpected runtime error renders the pipeline [`Unrechable`].
    Running,

    /// Graceful shutdown in progress.
    ///
    /// In this state, the pipeline finishes any ongoing data processing,
    /// produces final outputs, shuts down input/output connectors and terminates.
    ///
    /// The pipeline remains in this state until:
    ///
    /// 1. Shutdown completes successfully; transitions to the [`Shutdown`] state.
    /// 2. A pre-defined timeout has passed.  The manager performs forced shutdown of
    ///    the pipeline; returns to the [`Shutdown`] state.
    /// 3. The user cancels the pipeline by invoking the `/shutdown` endpoint.
    ///    The manager performs forced shutdown of the pipeline, returns to the
    ///    [`Shutdown`] state.
    ShuttingDown,

    /// The pipeline encountered a fatal error during initialization or at runtime.
    ///
    /// In this state, the pipeline's HTTP server is still running.  The manager is
    /// able to query the status of the pipeline, but the query engine has been shut
    /// down and no data is being processed.
    ///
    /// This state is not exposed to the user, since upon encountering the failure,
    /// the manager performs a forced shutdown of the pipeline and transitions to
    /// the [`Shutdown`] state.
    ///
    /// The pipeline remains in this state until:
    /// 1. The manager performs a forced shutdown of the pipeline; transitions to the
    ///    [`Shutdown`] state.
    /// 3. An unexpected runtime error renders the pipeline [`Unrechable`].
    Failed {
        error: ErrorResponse,
    },

    /// An unexpected runtime error renders the pipeline unreachable.
    ///
    /// This can be caused by a network or hardware failure, software panic, etc. 
    ///
    /// The pipeline remains in this state until:
    /// 1. The manager performs a forced shutdown of the pipeline; transitions to the
    ///    [`Shutdown`] state.
    /// 2. The transient error condition clears and the pipeline becomes reachable
    ///    again.
    Unreachable,
}
```

cc: @gz , @lalithsuresh

**Comment by gz** [comment]
> (/deploy, /shutodwn, /pause, /start) return instantly after forwarding the request to the pipeline, without waiting for the request to complete. 

Maybe it has to be like this because of some technical reason, but if not, why can't the endpoint be synchronous? e.g., I don't see the advantage of it being async there isn't much I can with the pipeline do (except polling stats to wait till it's up/down etc.).

```
    /// The pipeline remains in this state until:
    /// 1. The manager performs a forced shutdown of the pipeline; transitions to the
    ///    [`Shutdown`] state.
    /// 2. The transient error condition clears and the pipeline becomes reachable
    ///    again.
    Unreachable,
```
Wouldn't it transition to the `Failed` state instead of Shutdown coming from Unreachable? It's not clear how recovery from Unreachable will look like but I guess that can be defined later.

**Comment by lalithsuresh** [comment]
* I assume the local runner is doing the polling?
* In the cloud FF, the pipeline manager might not be the one deploying pipelines or managing their lifecycle.
* It sounds like we're obligating a manager in many instances to act within a timeout, which is not practical in a distributed system (e.g., failures? having multiple managers? partitions?). I'd suggest rewording such instances to `The manager forcibly shuts down the pipeline due to a timeout or in response to a user requested cancellation via the /shutdown endpoint`. 
* `Provisioning` -> `Deploying` --- so that it is consistent with the API operation used (`/deploy`).
*  `... error renders the pipeline [`Unrechable`]. `: typo
* All states except `Start` and `Pause` have the intermediate states in continuous tense. Are `start` and `pause` expected to be blocking operations? What does the manager see if it issues the call, fails before it completes, and comes back online?

**Comment by ryzhyk** [comment]
> > (/deploy, /shutodwn, /pause, /start) return instantly after forwarding the request to the pipeline, without waiting for the request to complete.
> 
> Maybe it has to be like this because of some technical reason, but if not, why can't the endpoint be synchronous? e.g., I don't see the advantage of it being async there isn't much I can with the pipeline do (except polling stats to wait till it's up/down etc.).

`/pause` and `/start` are in fact synchronous in the sense that if they returned `Ok`, the pipeline is paused or running respectively and the UI can safely report this status to the user.  After this the UI should keep polling the status because it can change due to an API call or a UI action from a different browser window.

`/deploy` and `/shutdown` can take a (relatively) long time. So the best course of action for the UI is to make the call and start polling, which should actually reduce the number of special cases in the UI, since you don't need to worry about what happens if the user navigates away from the pipelines page while the operation is running, closes the browser, the manager restarts etc. In all these cases the UI should always go to the manager as the source of ground truth about the state of the pipeline.

> ```
>     /// The pipeline remains in this state until:
>     /// 1. The manager performs a forced shutdown of the pipeline; transitions to the
>     ///    [`Shutdown`] state.
>     /// 2. The transient error condition clears and the pipeline becomes reachable
>     ///    again.
>     Unreachable,
> ```
> 
> Wouldn't it transition to the `Failed` state instead of Shutdown coming from Unreachable? It's not clear how recovery from Unreachable will look like but I guess that can be defined later.

Unreachable state can be caused by a transient network error or temporary CPU/IO overload.  Once the condition clears, the pipeline can be back to any of the other states, including Paused or Running.   The only way to recover from Unreachable state (other than waiting for it to clear by itself) is the "forced shutdown" action, which basically kills the process.

**Comment by gz** [comment]
> /pause and /start are in fact synchronous in the sense that if they returned Ok, the pipeline is paused or running respectively and the UI can safely report this status to the user. After this the UI should keep polling the status because it can change due to an API call or a UI action from a different browser window.

> /deploy and /shutdown can take a (relatively) long time. So the best course of action for the UI is to make the call and start polling, which should actually reduce the number of special cases in the UI, since you don't need to worry about what happens if the user navigates away from the pipelines page while the operation is running, closes the browser, the manager restarts etc. In all these cases the UI should always go to the manager as the source of ground truth about the state of the pipeline.

I'm not too worried about the UI since both ways will work; (the UI can ideally just read/update the status through the PipelineDesc whenever it gets queried)... but for any API user it means  POSTing an action and then making another series of GET calls until the status changes (which means some level of extra scripting).. So I don't really see why the POST can't just return when there is a status update which avoids the extra poll calls? Especially because I'm not sure what one can do e.g., I can't send start anyways before deploy completes etc.

**Comment by ryzhyk** [comment]
> * I assume the local runner is doing the polling?

Yes. I am still working out the separation of concerns between the manager and form-factor-specific runners.

>     * In the cloud FF, the pipeline manager might not be the one deploying pipelines or managing their lifecycle.

The manager _delegates_ specific tasks to other components (k8s or resource manager), but it is ultimately in charge.  For example the provisioning step consists of the pipeline manager writing desired state to k8s (or maybe the resource manager reads it directly from Postgres).  It is then the job of k8s to instantiate the pod and record its location somewhere.  The manager waits for this event and then moves the state machine to the next state (Initializing) and starts pollint its `/stats` endpoint to determine when initialization completes or fails.

>     * It sounds like we're obligating a manager in many instances to act within a timeout, which is not practical in a distributed system (e.g., failures? having multiple managers? partitions?). I'd suggest rewording such instances to `The manager forcibly shuts down the pipeline. For example, due to a timeout or in response to a user requested cancellation via the /shutdown endpoint`.

The assumption here is that the manager has a way to force-kill the pipeline, e.g., by changing desired state in k8s, even if the pipeline was actually healthy but unresponsive due to network partitioning.

>     * `Provisioning` -> `Deploying` --- so that it is consistent with the API operation used (`/deploy`).

The `deploy` operation consists of two steps: (1) Provisioning (i.e., creating the pod and starting the pipeline process within it), and (2) initializing the pipeline (e.g., connecting to Kafka topics).

>     * `... error renders the pipeline [`Unrechable`]. `: typo
> 
>     * All states except `Start` and `Pause` have the intermediate states in continuous tense. Since `start` and `pause` are not blocking operations, what does the manager see if it issues the call, fails before the call completes, and comes back online?

The nice thing about this design is that such situations don't pose any problems.  The server simply polls the state of the pipeline, as it does during normal operation (but probably forcing the initial poll early on).

**Comment by ryzhyk** [comment]
> > /pause and /start are in fact synchronous in the sense that if they returned Ok, the pipeline is paused or running respectively and the UI can safely report this status to the user. After this the UI should keep polling the status because it can change due to an API call or a UI action from a different browser window.
> 
> > /deploy and /shutdown can take a (relatively) long time. So the best course of action for the UI is to make the call and start polling, which should actually reduce the number of special cases in the UI, since you don't need to worry about what happens if the user navigates away from the pipelines page while the operation is running, closes the browser, the manager restarts etc. In all these cases the UI should always go to the manager as the source of ground truth about the state of the pipeline.
> 
> I'm not too worried about the UI since both ways will work; (the UI can ideally just read/update the status through the PipelineDesc whenever it gets queried)... but for any API user it means POSTing an action and then making another series of GET calls until the status changes (which means some level of extra scripting).. So I don't really see why the POST can't just return when there is a status update which avoids the extra poll calls? Especially because I'm not sure what one can do e.g., I can't send start anyways before deploy completes etc.

I don't think there's any different between UI and API in this case. The API user can safely assume that if a `/pause` call succeeded, the pipeline has been paused.  For long-running actions I suspect that a good API actually should not block.  One can build a blocking wrapper + timeout around it if needed.

The internal implementation also becomes much simpler if we eliminate long-running API actions, since locking and concurrency becomes very tricky to reason about for such calls.

**Comment by ryzhyk** [comment]
Another way to look at this: I am discovering that maintaining this state machine is tricky enough even with a minimal API. The main problem is race conditions between parallel API calls.  The goal here is to have an architecture that allows resolving such races by going to the pipeline as the source of ground truth and by careful synchronization when this is not possible (because the pipeline is not reachable).  If we can nail that, then more high-level APIs should be easy to build on top.

**Comment by gz** [comment]
For a future version of this, we should maybe have a state machine for each attached connector, so if e.g., one kafka connector fails the dbsp program will keep running.

**Comment by ryzhyk** [comment]
That's already the case. During initialization, the entire pipeline fails if one of the connectors fails to initialize. But at runtime a fatal error in one connector doesn't terminate the pipeline.

**Comment by gz** [comment]
Cool! We don't show individual error states on attached connectors (or can clear/try to reinitialize them) in the UI yet. Let's discuss offline at some point how to do that.

**Comment by ryzhyk** [comment]
No, at the moment pipeline configuration doesn't change at runtime (http connectors are an implementation detail that is not exposed through the API).
